src/internal/core.rs

```diff
@@ -3,55 +3,57 @@
 //! Core model and functions
 //! to write a functional PubGrub algorithm.
 
-use std::error::Error;
+use std::collections::HashSet as Set;
 use std::sync::Arc;
 
 use crate::error::PubGrubError;
 use crate::internal::arena::Arena;
-use crate::internal::incompatibility::{IncompId, Incompatibility, Relation};
+use crate::internal::incompatibility::{Incompatibility, Relation};
 use crate::internal::partial_solution::SatisfierSearch::{
     DifferentDecisionLevels, SameDecisionLevels,
 };
 use crate::internal::partial_solution::{DecisionLevel, PartialSolution};
 use crate::internal::small_vec::SmallVec;
-use crate::package::Package;
 use crate::report::DerivationTree;
-use crate::type_aliases::{DependencyConstraints, Map, Set};
+use crate::solver::DependencyProvider;
+use crate::type_aliases::{DependencyConstraints, IncompDpId, Map, V};
 use crate::version_set::VersionSet;
 
 /// Current state of the PubGrub algorithm.
 #[derive(Clone)]
-pub struct State<P: Package, VS: VersionSet, Priority: Ord + Clone> {
-    root_package: P,
-    root_version: VS::V,
+pub struct State<DP: DependencyProvider> {
+    root_package: DP::P,
+    root_version: V<DP>,
 
-    incompatibilities: Map<P, Vec<IncompId<P, VS>>>,
+    #[allow(clippy::type_complexity)]
+    incompatibilities: Map<DP::P, Vec<IncompDpId<DP>>>,
 
     /// Store the ids of incompatibilities that are already contradicted.
     /// For each one keep track of the decision level when it was found to be contradicted.
     /// These will stay contradicted until we have backtracked beyond its associated decision level.
-    contradicted_incompatibilities: Map<IncompId<P, VS>, DecisionLevel>,
+    contradicted_incompatibilities: Map<IncompDpId<DP>, DecisionLevel>,
 
     /// All incompatibilities expressing dependencies,
     /// with common dependents merged.
-    merged_dependencies: Map<(P, P), SmallVec<IncompId<P, VS>>>,
+    #[allow(clippy::type_complexity)]
+    merged_dependencies: Map<(DP::P, DP::P), SmallVec<IncompDpId<DP>>>,
 
     /// Partial solution.
     /// TODO: remove pub.
-    pub partial_solution: PartialSolution<P, VS, Priority>,
+    pub partial_solution: PartialSolution<DP>,
 
     /// The store is the reference storage for all incompatibilities.
-    pub incompatibility_store: Arena<Incompatibility<P, VS>>,
+    pub incompatibility_store: Arena<Incompatibility<DP::P, DP::VS>>,
 
     /// This is a stack of work to be done in `unit_propagation`.
     /// It can definitely be a local variable to that method, but
     /// this way we can reuse the same allocation for better performance.
-    unit_propagation_buffer: SmallVec<P>,
+    unit_propagation_buffer: SmallVec<DP::P>,
 }
 
-impl<P: Package, VS: VersionSet, Priority: Ord + Clone> State<P, VS, Priority> {
+impl<DP: DependencyProvider> State<DP> {
```

Comment:
We currently don't a `DependencyProvider`, but we do have a `State<PubGrubPackage, Range<Version>, PubGrubPriority>` (https://github.com/astral-sh/uv/blob/3799862f5de388c49fe7a4a1417213861aff7746/crates/uv-resolver/src/resolver/mod.rs#L242)

Reply:
I see two ugly ways to fix this. But I'm open to other suggestions:
- `pubgrub-rs` eats the ugliness, and we only use `DP` at the surface level. `State` and below have a large list of generic types, each of which is bound to match their `DP` associated type.
- `uv` eats the ugliness, you have some type that implements `DependencyProvider` with `unimplemented!()` for all required methods. As such, `State` can retrieve the associated types and no one needs to instantiate an instance.

I guess there's another option:
- we stabilize a `AsincDependencyProvider` and you switch to using our stable interface.

Thoughts?

Reply:
If uv's solver is heavily modified to fit the async needs, I wouldn't try to merge uv and pubgrub-rs solvers until more work is done on async in pubgrub-rs. Especially while uv is still in early days and more divergence may occur. Is this a priority of yours @konstin to re-depend on pubgrub-rs asap?

Reply:
I like not using a dependency provider, running the main loop ourselves makes the code straightforward (no callbacks) and we can hook in at many places.

Reply:
That's good to know!

Reply:
@konstin:
- Just to confirm my understanding `State` is not part of our public API. Is that correct? If we accidentally made it public I would very much like to know.
- Similarly, how painful would it be for you to make a fake `DependencyProvider` in `uv` for `State` to read its associated types out of? It would not have to implement any of the methods.

Reply:
Yes, we made `State` public (https://github.com/pubgrub-rs/pubgrub/compare/dev...astral-sh:pubgrub:main#diff-cbd16a2b1fdb7e2f39b676831dfaee5250000b403a9f1fcfcbfede424d7cbf73L78-R79)

If it's just for the typings it's fine to implement a `DependencyProvider`.

Reply:
That's probably a discussion to be had separate from this issue, but how hard would it be to make state or a state wrapper public, to officially support using pubgrub from your own solver loop instead of using pubgrub's? As in, making enough public that you can implement `resolve` in user code?

Reply:
From my point of view, I like having package APIs that offer different ease-of-use/control tradeoffs. First level is friendliest and easiest to use but don't offer much control. Up to now we only had 2 levels, first level is dependency provider already done, second level is you customize the dep provider. It seemed the right balance without sacrificing correctness. In your case, you need an additional level of control, where even more correctness is in the hand of the user. Anyway just my 2 cents. It might be worth exploring.